herddb-core/src/main/java/herddb/core/SimpleDataScanner.java

```diff
@@ -49,6 +51,11 @@ public SimpleDataScanner(Transaction transaction, MaterializedRecordSet recordSe
 
     @Override
     public void close() throws DataScannerException {
+        if (closed.compareAndSet(false, true)) {
+            if (transaction != null) {
+                transaction.decreaseRefCount();
+            }
+        }
         finished = true;
```

Comment:
Can't be added the whole closing method under close.compareAndSet?
And practically speaking: could we just use an atomiboolean for both "closed" and  "finished" (and "fast closing" the scanner and result set upon recordSet finishing?)

Reply:
"fast closing" the scanner and result set upon recordSet finishing? -> this may be lead to unpredictable behaviour.

We are closing a "finished" resultset in ServerSideConnectionPeer and this is part of the "protocol": if there are no more records the ResultSet is closed, and the client won't send the "CLOSE SCANNER" message (this is the most common case for findByPk for instance)

Reply:
btw I have moved the block inside the critical section